src/main/java/net/neoforged/neoforge/common/conditions/ConditionalOps.java

```diff
@@ -61,123 +68,105 @@ public static <T> Codec<List<T>> decodeListWithElementConditionsAndConsumeIndex(
                 Function.identity()));
     }
 
-    public static <T> Decoder<Optional<T>> createConditionalDecoder(final Decoder<T> ownerCodec) {
-        return createConditionalDecoder(ownerCodec, "conditions");
+    /**
+     * @see #createConditionalDecoder(Decoder, String)
+     */
+    public static <T> Decoder<Optional<T>> createConditionalDecoder(final Decoder<T> ownerDecoder) {
+        return createConditionalDecoder(ownerDecoder, "conditions");
     }
 
-    public static <T> Decoder<Optional<T>> createConditionalDecoder(final Decoder<T> ownerCodec, String conditionalsKey) {
-        return new ConditionalDecoder<T>(
-                conditionalsKey,
-                ICondition.LIST_CODEC,
-                retrieveContext().codec(),
-                ownerCodec).map(o -> o.map(WithConditions::carrier));
+    /**
+     * Creates a conditional decoder.
+     * If the inner object (of type {@code T}) serializes as a map, then conditions are checked.
+     * Otherwise, conditions are not looked for.
+     * In other words, the deserializer does not assume that the inner object is a map.
+     */
+    public static <T> Decoder<Optional<T>> createConditionalDecoder(final Decoder<T> ownerDecoder, String conditionalsKey) {
+        return new ConditionalDecoder<>(conditionalsKey, ICondition.LIST_CODEC, retrieveContext().codec(), ownerDecoder)
+                .map(r -> r.map(WithConditions::carrier));
     }
 
-    public static <T> MapCodec<Optional<WithConditions<T>>> createConditionalCodecWithConditions(final Codec<T> ownerCodec) {
+    /**
+     * @see #createConditionalCodecWithConditions(Codec, String)
+     */
+    public static <T> Codec<Optional<WithConditions<T>>> createConditionalCodecWithConditions(final Codec<T> ownerCodec) {
         return createConditionalCodecWithConditions(ownerCodec, "conditions");
     }
 
-    public static <T> MapCodec<Optional<WithConditions<T>>> createConditionalCodecWithConditions(final Codec<T> ownerCodec, String conditionalsKey) {
-        return new ConditionalCodec<T>(
-                conditionalsKey,
-                ICondition.LIST_CODEC,
-                retrieveContext().codec(),
-                ownerCodec);
+    /**
+     * Creates a conditional codec. The codec assumes that the inner object (of type {@code T}) serializes as a map.
+     */
+    public static <T> Codec<Optional<WithConditions<T>>> createConditionalCodecWithConditions(final Codec<T> ownerCodec, String conditionalsKey) {
```

Comment:
Probably worth adding a comment somewhere here noting the use of a codec created by this in anywhere that is delegated to by a KeyDispatchCodec may have differing behaviour, on decoding, than the original codec, since we don't want someone accidentally using it that way in a neo patch and breaking vanilla compat again in the future

Reply:
Because it's not a `MapCodec` you mean?

Reply:
Yeah, cause it's not a MapCodecCodec and KeyDispatchCodec is funky. As long as it's documented it should be fine

Reply:
Added a bit of javadoc for that in 33eabb0.